discovery/swarmkit/swarmkit.go

```diff
@@ -0,0 +1,160 @@
+package swarmkit
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/discovery"
+	"github.com/docker/swarm/cluster"
+)
+
+const defaultAgentPort = "2375"
+
+// Discovery is exported.
+type Discovery struct {
+	heartbeat  time.Duration
+	ttl        time.Duration
+	httpClient *http.Client
+	apiClient  *client.Client
+	uri        string
+	token      string
+	agentPort  int
+}
+
+func init() {
+	Init()
+}
+
+// Init is exported.
+func Init() {
+	discovery.Register("swarmkit", &Discovery{})
+}
+
+func (d *Discovery) SetAgentPort(discoveryEnginePort int) {
+	d.agentPort = discoveryEnginePort
+}
+
+// Initialize initializes the discovery with the SwarmKit manager's address
+func (d *Discovery) Initialize(uri string, heartbeat time.Duration, ttl time.Duration, _ map[string]string) error {
+	// the uri may be a manager address or a docker socket
+	if uri == "" {
+		return errors.New("Swarm Mode manager address not provided")
+	}
+	d.uri = uri
+	d.heartbeat = heartbeat
+	d.ttl = ttl
+
+	// create a client to connect with the Docker API
+	httpClient, _, err := cluster.NewHTTPClientTimeout("tcp://"+uri, nil, time.Duration(30*time.Second), nil)
+	if err != nil {
+		return err
+	}
+	d.httpClient = httpClient
+
+	apiClient, err := client.NewClient("tcp://"+uri, "", d.httpClient, nil)
```

Comment:
Shouldn't we do a version negotiation here?

Reply:
You're right. Added a call to `NegotiateAPIVersion`.